docs/src/main/sphinx/develop/tests.md

```diff
@@ -0,0 +1,302 @@
+  necessary to verify this integration.
+
+### Avoid creating bespoke testing abstractions
+
+unless absolutely necessary. Examples of this are:
+
+* Custom dispatch frameworks for parallelizing test execution
+* Test-specific assertion frameworks
+* Custom parameterized test frameworks
+
+### Avoid data providers and parametric tests
+
+* Most existing data providers are either trivially small (i.e., a single
+  boolean), or massive combinatorial, indiscriminate, data generation
+* Prefer to just invoke tests cases explicitly for trivial cases like a boolean
+  parameter
+* For small datasets, use a “for-each item in an inline list”
+* For larger datasets, use a type safe enum class
+    * Read “Focus on high value tests”, and reconsider if the test is needed at
+      all
+    * For large test datasets, you must talk to the Testing Lead. We are
+      collecting examples of cases where this is needed, and places where they
+      were not, so we can write better guidelines.
+* Avoid multiple independent data providers in a test
+    * This can be multiple nested for loops or multiple data provider parameters
+
+### Avoid writing stateful test classes
+
+ if possible. They are harder to maintain and parallelize. Bugs in one test can
+ leak to other tests and make debugging harder.
+
+### Do not try to manage memory
+
+Do not null out fields in @After methods to “free memory” JUnit, unlike TestNG,
+has proper test life cycle management, so it is safe to assign memory intensive
+objects to final fields, as the class will be dereferenced after the test run.
+
+### Use simple resource initialization
+
+Prefer resource initialization in constructors and tear them down in @After
+methods if necessary This, along with not nulling fields, allows the fields to
+be final and behave like any Closeable class would work in normal Java code
+Consider using the Guava Closer class to simplify cleanup
+
+### Keep test setup and teardown simple
+
+Avoid the @Before/After each test method style of setup and teardown
+
+* Prefer try-with-resources if natural
+* If necessary, use a shared initialization or cleanup method that is explicitly
+  called
+* If you have a test that benefits from @Before/After methods, please talk to
+  the Testing Lead, as we are working to refine the guidance here.
+
+### Ensure testability of new plugin and connector features
+
+New plugin/connector features should be testable using one of the testing
+plugins (e.g., memory or null). There are existing features only tested in
+plugins in Hive, and over time we expect coverage using the testing plugins
+
+### Keep focus on plugin and connector tests
+
+For plugins/connectors, focus on the code unique to the connector, and do not
+re-test engine features Plugins should be focused on correctness of the SPI
+implementation and compatibility with external systems.
+
+### Avoid flaky tests
+
+Flaky tests can be marked with the @Flaky annotation temporarily to improve CI
```

Comment:
AFAIK this doesn't do anything with JUnit, does it?

If so, we either leave it out of the guidelines, or specifically call out that this works for TestNG tests only

Reply:
Sounds good .. I still have to update all this to be a guideline on what to do  .. vs the migration project

Reply:
Should we remove this bullet @martint ?

Reply:
Reworded as well btw.